rts/Lua/LuaSyncedRead.cpp

```diff
@@ -3652,14 +3626,80 @@ int LuaSyncedRead::GetProjectilesInRectangle(lua_State* L)
 			if (!LuaUtils::IsProjectileVisible(L, pro))
 				continue;
 
-			lua_pushnumber(L, pro->id);
-			lua_rawseti(L, -2, arrayIndex++);
+			lua_pushnumber(L, static_cast<lua_Number>(pro->id));
+			lua_rawseti(L, -2, static_cast<int>(arrayIndex++));
 		}
 	}
+}
 
+/***
+ *
+ * @function Spring.GetProjectilesInRectangle
+ * @param xmin number
+ * @param zmin number
+ * @param xmax number
+ * @param zmax number
+ * @param excludeWeaponProjectiles boolean? (Default: false)
+ * @param excludePieceProjectiles boolean? (Default: false)
+ * @return number[] projectileIDs
+ */
+int LuaSyncedRead::GetProjectilesInRectangle(lua_State* L)
+{
+	const float xmin = luaL_checkfloat(L, 1);
+	const float zmin = luaL_checkfloat(L, 2);
+	const float xmax = luaL_checkfloat(L, 3);
+	const float zmax = luaL_checkfloat(L, 4);
+
+	const bool excludeWeaponProjectiles = luaL_optboolean(L, 5, false);
+	const bool excludePieceProjectiles = luaL_optboolean(L, 6, false);
+
+	const float3 mins(xmin, 0.0f, zmin);
+	const float3 maxs(xmax, 0.0f, zmax);
+
+	QuadFieldQuery qfQuery;
+	quadField.GetProjectilesExact(qfQuery, mins, maxs);
+	GetProjectilesLuaTable(L, *qfQuery.projectiles, excludeWeaponProjectiles, excludePieceProjectiles);
 	return 1;
 }
 
+/***
+ *
+ * @function Spring.GetProjectilesInSphere
+ * @param x number
+ * @param y number
+ * @param z number
+ * @param radius number
+ * @param excludeWeaponProjectiles boolean? (Default: false)
+ * @param excludePieceProjectiles boolean? (Default: false)
+ * @return number[] projectileIDs
+ */
+int LuaSyncedRead::GetProjectilesInSphere(lua_State* L)
+{
+	const float3 sphereCenter(luaL_checkfloat(L, 1), luaL_checkfloat(L, 2), luaL_checkfloat(L, 3));
+	const float radius = luaL_checkfloat(L, 4);
+	const float radSqr = radius * radius;
+
+	const bool excludeWeaponProjectiles = luaL_optboolean(L, 5, false);
+	const bool excludePieceProjectiles = luaL_optboolean(L, 6, false);
+
+	const float3 mins(sphereCenter.x - radius, 0.0f, sphereCenter.z - radius);
+	const float3 maxs(sphereCenter.x + radius, 0.0f, sphereCenter.z + radius);
+
+	QuadFieldQuery qfQuery;
+	quadField.GetProjectilesExact(qfQuery, mins, maxs);
+
+	std::vector<CProjectile *> inSphereProjectiles;
+	inSphereProjectiles.reserve(qfQuery.projectiles->size());
+	for (unsigned int i = 0; i < qfQuery.projectiles->size(); i++) {
+		const auto projectile = (*qfQuery.projectiles)[i];
+		const float3 projectilePos(projectile->pos.x, projectile->pos.y, projectile->pos.z);
+		if (projectilePos.SqDistance(sphereCenter) <= radSqr) {
```

Comment:
`projectile->pos` should already be a `float3`?
```suggestion
		if (projectile->pos.SqDistance(sphereCenter) <= radSqr) {
```

Reply:
Also maybe the manual distance check is not needed at all since it looks like there's a version of GetProjectilesExact that seems to work on a sphere?
https://github.com/beyond-all-reason/spring/blob/95d591b7c91f26313b58187692bd4485b39cb050/rts/Sim/Misc/QuadField.cpp#L811

Reply:
Ty. Missed that signature.